common/src/main/java/io/pravega/common/util/SortedDeque.java

```diff
@@ -0,0 +1,269 @@
+        int size = size();
+        while (itemIndex++ < size) {
+            int t = (this.tail - 1) & this.lastSlotIndex;
+            this.elements[t] = null;
+            this.tail = t;
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns the first item in the Deque, without removing it.
+     *
+     * @return The first item in the Deque, or null if Deque is empty.
+     */
+    public E peekFirst() {
+        return (E) this.elements[this.head];
+    }
+
+    /**
+     * Returns the last item in the Deque, without removing it.
+     *
+     * @return The last item in the Deque, or null if Deque is empty.
+     */
+    public E peekLast() {
+        return (E) this.elements[(this.tail - 1) & this.lastSlotIndex];
+    }
+
+    /**
+     * Clears the entire Deque.
+     */
+    public void clear() {
+        if (this.head != this.tail) {
+            do {
+                this.elements[this.head] = null;
+                this.head = (this.head + 1) & this.lastSlotIndex;
+            } while (this.head != this.tail);
+            this.head = this.tail = 0;
+        }
+    }
+
+    /**
+     * Returns the number of elements in this Deque.
+     *
+     * @return The result.
+     */
+    public int size() {
+        return (this.tail - this.head) & this.lastSlotIndex;
+    }
+
+    /**
+     * Returns true if this Deque contains no elements.
+     *
+     * @return The result.
+     */
+    public boolean isEmpty() {
+        return this.head == this.tail;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("Size = %d, Capacity = %d", size(), this.elements.length);
+    }
+
+    //endregion
+
+    //region Helpers
+
+    /**
+     * Doubles the capacity of the Deque by creating a new array double of the current size and copying all elements into
+     * it.
+     */
+    private void doubleCapacity() {
+        assert this.head == this.tail : "not at capacity";
+        int oldHead = this.head;
+        int oldLength = this.elements.length;
+        int r = oldLength - oldHead; // number of elements to the right of oldHead
+        int newCapacity = oldLength << 1;
+        Preconditions.checkState(newCapacity >= 0, "Deque too large.");
+
+        SortedIndex.IndexEntry[] newElements = new SortedIndex.IndexEntry[newCapacity];
+        System.arraycopy(this.elements, oldHead, newElements, 0, r);
+        System.arraycopy(this.elements, 0, newElements, r, oldHead);
+        this.elements = newElements;
+        this.lastSlotIndex = this.elements.length - 1;
+        this.head = 0;
+        this.tail = oldLength;
+    }
+
+    /**
+     * Performs a binary search in the Deque for an item whose key matches the given one.
+     *
+     * @param key The key to search.
+     * @return The index in the element in the Deque (NOT the index in the base array), or -1 if no such element exists.
+     */
+    private int binarySearch(long key) {
```

Comment:
I looked at that, but it's not what I need. `head` can be after `tail` in which case my array wraps around. The JRE's binary search does not handle that case, hence the need for my custom method.